src/bap_recipe.ml

```diff
@@ -89,60 +110,111 @@ let parse_recipe str =
 
 let parse_descr str = Ok str
 
-let (/) = Filename.concat
-
-let make_path path name =
-  path / sprintf "%s.recipe" name
+let get_recipe w = input w.main parse_recipe w
+let get_descr w = match input "descr" parse_descr w with
+  | Error (Missing_entry _) -> Ok "Description was not provided"
+  | x -> x
 
 let subst_string f str =
   let buf = Buffer.create (String.length str) in
   Buffer.add_substitute buf f str;
   Buffer.contents buf
 
-let make_subst files pars vars arg =
-  match Map.find files arg with
-  | Some x -> x
-  | None ->  match Map.find vars arg with
+let make_subst root pars vars arg =
+  if arg = "prefix" then root.path
+  else match Map.find vars arg with
     | Some x -> x
     | None -> List.find_map pars ~f:(fun {name; defl} ->
         if name = arg then Some defl else None) |> function
               | Some x -> x
-              | None -> try Sys.getenv arg with
-                | Not_found -> raise (Bad_substitution arg)
+              | None -> raise (Bad_substitution arg)
 
-let apply_subst_exn files pars vars opts =
-  let subst = make_subst files pars vars in
+let apply_subst_exn root pars vars opts =
+  let subst = make_subst root pars vars in
   List.map opts ~f:(fun (name,args) ->
       "--"^name,List.map args ~f:(subst_string subst))
 
-let apply_subst files pars vars opts =
-  try Ok (apply_subst_exn files pars vars opts)
+let apply_subst root pars vars opts =
+  try Ok (apply_subst_exn root pars vars opts)
   with Bad_substitution s -> Error (Bad_subst s)
 
-let linearize  =
-  List.concat_map ~f:(fun (name,xs) -> name :: xs)
+let linearize =
+  List.map ~f:(fun (name,xs) -> match xs with
+      | [] -> name
+      | xs -> String.concat ~sep:"=" [
+          name;
+          String.concat ~sep:"," xs
+        ])
+
+let rng = Caml.Random.State.make_self_init ()
+
+let mkdtemp ?(mode=0o0700) ?tmp_dir ?(prefix="") ?(suffix="") () =
+  let genname () = Uuidm.v4_gen rng () |> Uuidm.to_string in
+  let rec create name =
+    let tmp = match tmp_dir with
+      | None -> Filename.get_temp_dir_name ()
+      | Some tmp -> tmp in
+    let path =
+      String.concat [tmp; Filename.dir_sep; prefix; name; suffix] in
+    match Unix.mkdir path mode with
```

Comment:
Will `mkdir`  create directories recursively?

Reply:
No, it won't. Neither it should.